src/actions/cart-manager.ts

```diff
@@ -1,92 +1,60 @@
-    return availableAmount - desiredAmount;
+    const availableQuantity = productInDB.quantity - getQuantityOfItemInMainCart(productId);
+    if (availableQuantity < desiredQuantity) {
+        throw new Error('Quantity not enough!');
+    }
+    mainCart.get(productId).set(userId, desiredQuantity);
+    return availableQuantity - desiredQuantity;
 };
 
 export const removeFromCart = async (userId: string, productId: string) => {
-    const product = await retrieveSpecificProduct(productId);
-    if (!product)
-        throw new Error('Item not found!');
-    const productInMainCart = findProductInMainCart(productId);
-    if (productInMainCart) {
-        removeUserFromItemInMainCart(userId, productInMainCart);
-        return product.quantity - getAmountOfItemInMainCart(productInMainCart);
-    } else
-        throw new Error('Item not found!');
-};
-
-const findProductInMainCart: (id: string) => any = (id: string) => {
-    let foundItem;
-    mainCart.forEach(cartItem => {
-        if (cartItem.id === id)
-            foundItem = cartItem;
-    });
-    return foundItem ? foundItem : null;
-};
-
-const getAmountOfItemInMainCart = (cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
-    let amountInCarts = 0;
-    cartItem.users.forEach(user => amountInCarts += user.amount);
-    return amountInCarts;
-};
-
-const updateProductInMainCart = (userId: string, amount: number, cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
-    let userFound = false;
-    cartItem.users.forEach(user => {
-        if (user.userId === userId) {
-            user.amount += amount;
-            userFound = true;
-        }
-    });
-    if (!userFound) {
-        const newUser = {userId, amount};
-        cartItem.users.push(newUser);
+    const productInDB = await retrieveSpecificProduct(productId);
+    if (!productInDB) {
+        throw new Error('Product not found!');
     }
-};
-
-const removeUserFromItemInMainCart = (userId: string, cartItem: { id: string, users: { userId: string, amount: number }[] }) => {
-    const newUsers = cartItem.users.filter(user => user.userId !== userId);
-    if (newUsers.length === cartItem.users.length) {
-        throw new Error('Item not in user cart')
-    } else {
-        cartItem.users = newUsers;
+    if (!mainCart.has(productId) || !mainCart.get(productId).has(userId)) {
+        throw new Error('Product not in user cart!');
     }
+    mainCart.get(productId).delete(userId);
+    return productInDB.quantity - getQuantityOfItemInMainCart(productId);
 };
 
 export const buildBuyListForUser = (userId: string) => {
     const buyList: { id: string; quantity: number; }[] = [];
-    mainCart.forEach(itemInMainCart => {
-        itemInMainCart.users.forEach(user => {
-            if (user.userId === userId)
-                buyList.push({id: itemInMainCart.id, quantity: user.amount});
-        });
-    });
+    mainCart.forEach(((usersMap, productId) => {
```

Comment:
use reduce

Reply:
are you sure it's possible to reduce a map? didn't find something like it